File: mediatorpattern.py
from abc import ABC, abstractmethod

class Colleague(ABC):
    @abstractmethod
    def place_bid(self, bid_amount):
        pass

    def receive_bid_notification(self, bid_amount):
        pass

    def get_name(self):
        pass

class Bidder(Colleague):
    def __init__(self, name, auction_mediator):
        self.name = name
        self.auction_mediator = auction_mediator
        self.auction_mediator.add_bidder(self)

    def place_bid(self, bid_amount):
        self.auction_mediator.place_bid(self, bid_amount)

    def receive_bid_notification(self, bid_amount):
        print('Bidder:', self.name, 'got notification that someone has put a bid of:', bid_amount)

    def get_name(self):
        return self.name
    
# this is a mediator interface
class AuctionMediator(ABC):
    @abstractmethod
    def add_bidder(self, bidder):
        pass

    @abstractmethod
    def place_bid(self, bidder, bid_amount):
        pass

class Auction(AuctionMediator):
    def __init__(self):
        self.colleagues = []

    def add_bidder(self, bidder):
        self.colleagues.append(bidder)

    def place_bid(self, bidder, bid_amount):
        for colleague in self.colleagues:
            if colleague.get_name() != bidder.get_name():
                colleague.receive_bid_notification(bid_amount)

if __name__ == '__main__':
    auction_mediator_obj = Auction()
    bidder_1 = Bidder('A', auction_mediator_obj)
    bidder_2 = Bidder('B', auction_mediator_obj)

    bidder_1.place_bid(2000)
    bidder_2.place_bid(3000)
    bidder_1.place_bid(3001)

File: mediatorpattern/mediatorpattern.js
// ===== Abstract Colleague =====
class Colleague {
  placeBid(bidAmount) {
    throw new Error('placeBid() must be implemented');
  }

  receiveBidNotification(bidAmount) {
    // Default empty implementation
  }

  getName() {
    throw new Error('getName() must be implemented');
  }
}

// ===== Concrete Colleague: Bidder =====
class Bidder extends Colleague {
  constructor(name, auctionMediator) {
    super();
    this.name = name;
    this.auctionMediator = auctionMediator;
    this.auctionMediator.addBidder(this);
  }

  placeBid(bidAmount) {
    this.auctionMediator.placeBid(this, bidAmount);
  }

  receiveBidNotification(bidAmount) {
    console.log(`Bidder: ${this.name} got notification that someone has put a bid of: ${bidAmount}`);
  }

  getName() {
    return this.name;
  }
}

// ===== Abstract Mediator =====
class AuctionMediator {
  addBidder(bidder) {
    throw new Error('addBidder() must be implemented');
  }

  placeBid(bidder, bidAmount) {
    throw new Error('placeBid() must be implemented');
  }
}

// ===== Concrete Mediator: Auction =====
class Auction extends AuctionMediator {
  constructor() {
    super();
    this.colleagues = [];
  }

  addBidder(bidder) {
    this.colleagues.push(bidder);
  }

  placeBid(bidder, bidAmount) {
    for (const colleague of this.colleagues) {
      if (colleague.getName() !== bidder.getName()) {
        colleague.receiveBidNotification(bidAmount);
      }
    }
  }
}

// ===== Client Code =====
const auctionMediatorObj = new Auction();
const bidder1 = new Bidder('A', auctionMediatorObj);
const bidder2 = new Bidder('B', auctionMediatorObj);

bidder1.placeBid(2000);
// Bidder: B got notification that someone has put a bid of: 2000

bidder2.placeBid(3000);
// Bidder: A got notification that someone has put a bid of: 3000

bidder1.placeBid(3001);
// Bidder: B got notification that someone has put a bid of: 3001
